### PR TITLE
allow list interpolations

### LINKED ISSUE #6989: Make frozen field configurable
# Bug Report

## Issue name
Freeze a stage in `dvc.yaml` right now, it is not possible by using the `params.yaml`.

## Description
`PARAM_FROZEN` which is defined in `dvc/schema.py` [here](https://github.com/iterative/dvc/blob/master/dvc/schema.py#L20) and [here](https://github.com/iterative/dvc/blob/master/dvc/schema.py#L87), accepts only bool values for now.

That means that I can do the following:
```yaml
# dvc.yaml
  train:
    cmd: PYTHONPATH='../../.' python3 training/run_train.py
    frozen: true
```

But I cannot have something like this:
```yaml
# params.yaml
model:
  freeze: false
```

```yaml
# dvc.yaml
  train:
    cmd: PYTHONPATH='../../.' python3 training/run_train.py
    frozen: ${model.freeze}
```
which gives me the following error:
```
ERROR: 'dvc.yaml' format error: extra keys not allowed @ data['stages']['train']['cmd']
```

### Reproduce
If you define any stage with the following format, you will get the error:
```yaml
# params.yaml
model:
  freeze: false
```

```yaml
# dvc.yaml
  train:
    cmd: PYTHONPATH='../../.' python3 training/run_train.py
    frozen: ${model.freeze}
```

### Expected
I would expect this `frozen` parameter to be configurable by defining `PARAM_FROZEN` in [here](https://github.com/iterative/dvc/blob/master/dvc/schema.py) in  a way like
`StageParams.PARAM_FROZEN: Any(str, bool)` instead of just `StageParams.PARAM_FROZEN: bool`.

I tried this locally and it solved my issue by the way.

### Environment information

**Output of `dvc doctor`:**

```console
$ dvc doctor
```
DVC version: 2.3.0 (pip)
---------------------------------
Platform: Python 3.8.8 on Linux-5.4.0-90-generic-x86_64-with-glibc2.10
Supports: http, https, ssh
Cache types: symlink
Cache directory: ext4 on /dev/sda
Caches: local
Remotes: ssh
Workspace directory: ext4 on /dev/nvme1n1p6
Repo: dvc, git

### LINKED ISSUE #7151: flatten `deps`  to facilitate more flexible templating
The templating feature currently admits little flexibility for stage keys (e.g. `deps`) that expect lists. As a result, `foreach` stages cannot be given a variable number of dependencies. Causing lists to be flattened would make templating more powerful. For example, you could give extra dependencies to certain templated stages:

```
stages:
  build:
    foreach:
      us:
        deps:
      uk:
        deps:
        - uk_data
    do:
      cmd: myscript ${key}
      deps:
      - myscript
      - ${item.deps}
```

I have implemented the change at https://github.com/itcarroll/dvc/blob/feature/flattened-deps/dvc/dependency/__init__.py#L52, by (non-recursive) list [flattening](https://github.com/itcarroll/dvc/blob/feature/flattened-deps/dvc/dependency/__init__.py#L21-L26) in `dvc.dependency.loads_from`. I am sure it's not the right way to do it, and probably has unintended consequences, but it demonstrates the concept (and no tests failed).

I don't know how templating responds to missing keys, but not having to include the empty list for `us` would be even more better.

### LINKED ISSUE #8171: Unpacking list templates outside `cmd`
For one project, I am running a foreach stage, where each step may have a different list of dependencies. I wanted to use templating to fill in the list of dependencies, but it seems for whatever I tried, the syntax was not supported. Below is a simple example that doesn't include the `foreach`:

```YAML
# params.yaml
those:
  - script.py
  - script2.py
```

I first tried

```YAML
# dvc.yaml
stages:
  hello_world:
    cmd: python script.py
    deps: 
      - ${those}
    outs:
      - hello.txt
```

Running `dvc repro` I get

```bash
>>> dvc repro
ERROR: failed to parse 'stages.hello_world.deps' in 'dvc.yaml':
Cannot interpolate data of type 'list'
```

This wouldn't have made that much sense to me anyways since it reads like I'm expanding a list within an entry of another list.

I also tried

```YAML
stages:
  hello_world:
    cmd: python script.py
    deps: ${those}
    outs:
      - hello.txt
```

and got

```bash
>>> dvc repro
'./dvc.yaml' validation failed.

expected a list, in stages -> hello_world -> deps, line 3, 
column 5
  2   hello_world:                                          
  3 │   cmd: python script.py                               
  4 │   deps: ${those} 
```

This one makes more sense since it reads similar to how the expansion in `foreach` stages works.

Please let me know
1. Is what I'm trying to achieve already supported? In that case the [templating documentation](https://dvc.org/doc/user-guide/project-structure/dvcyaml-files#templating) should be updated, since this behavior is not described anywhere.
2. If this is not currently supported, can this be supported in the future? The latter syntax would be a natural extension of the syntax for templating in `foreach` stages.

### PR DESCRIPTION
**Pull request by dberenbaum**
Opening for discussion. This approach does 2 things:

1. Ignores schema validation errors for all `${interpolated}` strings. Fixes #6989.
2. Allows interpolated values to be lists for any sections where lists are allowed (deps, outs, params, metrics, plots).

Discussed with @skshetry that another option here would be to revalidate the schema after interpolation, but the error handling there could be confusing.

The 2nd point above fixes #8171 since you can do:

```yaml
stages:
  hello_world:
    cmd: python script.py
    deps: ${those}
    outs:
      - hello.txt
```

It doesn't address #7151 because you still can't append to a list like:

```yaml
stages:
  build:
    foreach:
      us:
        deps:
      uk:
        deps:
        - uk_data
    do:
      cmd: myscript ${key}
      deps:
      - myscript
      - ${item.deps}
```

For that, we probably would need to introduce some new syntax.

### CODE REVIEW

**Comment by shcheklein** [comment]
> For that, we probably would need to introduce some new syntax.

Q: why would we need a new syntax?

**Comment by dberenbaum** [comment]
> You would just need to duplicate the shared dependencies inside each individual `${item.deps}`, right? I think it covers the use case

Do you mean like below? I think we could suggest it, but I'd like to keep open the issue because it feels limited for realistic use cases. For example, I have some list of data sources in `params.yaml` that needs to be reused as deps in multiple stages, but I also need to include the code dep for each stage.

```
stages:
  build:
    foreach:
      us:
        deps:
        - myscript
      uk:
        deps:
        - myscript
        - uk_data
    do:
      cmd: myscript ${key}
      deps:
      - ${item.deps}
```

> Q: why would we need a new syntax?

@shcheklein I should have said this PR doesn't fix that problem. That's all I meant.

File: dvc/parsing/interpolate.py [hunk 1]
```diff
@@ -203,6 +203,16 @@ def validate_value(value, key):
...
             return True
+        if isinstance(value, list) and key in [
+            "deps",
+            "outs",
+            "params",
+            "metrics",
+            "plots",
+        ]:
+            for e in value:
+                validate_value(e, key)
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply dvc/parsing/interpolate.py:214]: The rationale for why I did it this way was strictly simplicity, and because it follows what you already did for dict unpacking.

I'm open to re-validating but would need to take another look to figure out how to do it (feel free to try it). The concern about error handling was that the errors won't match what the users sees in `dvc.yaml` since this would be a sort of virtual copy with interpolated values expanded, right?

[the remark this answers is not in the document]
Comment (daavoo) [reply dvc/parsing/interpolate.py:214]: Makes sense. To clarify, I don't consider it a blocker.

[the remark this answers is not in the document]
Comment (skshetry) [reply dvc/parsing/interpolate.py:214]: > i.e. frozen must be bool

That way, you have to reimplement all of the schema validation. I don't really like what we do for `cmd` already. It gets harder too, eg: `outs` could be a list of dict with specific values.

**Comment by efiop** [comment]
@daavoo @skshetry What's the status here?

**Comment by dberenbaum** [comment]
Seems like there's no opposition to this. Let me fix the tests and I'll ask you to review one more time @daavoo @skshetry.

**Review by daavoo — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -53,7 +53,7 @@ dependencies = [
     "pydot>=1.2.4",
     "pygtrie>=2.3.2",
     "pyparsing>=2.4.7",
-    "requests>=2.22",
+    "requests>=2.28",
```
Comment (dberenbaum) [inline pyproject.toml:56]: Unrelated change to fix studio tests

**Comment by dberenbaum** [comment]
Getting some seemingly unrelated errors, but it should be ready for review.

File: pyproject.toml (hunk 2 above)
Comment (skshetry) [reply pyproject.toml:56]: Which one requires this update?

Comment (skshetry) [reply pyproject.toml:56]: This should be fixed in the `main` now (dependency update is not required). I suggest upgrading packages from time to time using:
```console
pip install --upgrade --upgrade-strategy eager -e ".[dev]"
```

This one is our dependency, but you won't run into mypy and other linting issue if you upgrade.

Comment (dberenbaum) [reply pyproject.toml]: Thanks @skshetry. Reverted. PTAL.

**Comment by dberenbaum** [comment]
ping @skshetry could you review when you have a chance or merge if you don't think it's needed?

File: tests/func/parsing/test_interpolated_entry.py [hunk 3]
```diff
@@ -315,3 +321,19 @@ def test_cmd_dict(tmp_dir, dvc, bool_config, list_config):
             }
         },
     )
+
+
+def test_interpolate_bool(tmp_dir, dvc):
```
Comment (skshetry) [inline tests/func/parsing/test_interpolated_entry.py:326]: We don't need this test. This is already tested:

https://github.com/iterative/dvc/blob/071a5b7ba4a545aa418e880127bc34f60fe0a3b0/tests/unit/test_context.py#L409

File: dvc/utils/strictyaml.py [hunk 4]
```diff
@@ -261,12 +264,25 @@ def validate(
...
     rev: Optional[str] = None,
 ) -> _T:
-    from voluptuous import MultipleInvalid
+    from voluptuous import MultipleInvalid, TypeInvalid
 
     try:
         return schema(data)
     except MultipleInvalid as exc:
-        raise YAMLValidationError(exc, path, text, rev=rev) from exc
+        # Drop errors for str interpolation values
```
Comment (skshetry) [inline dvc/utils/strictyaml.py:272]: I am not comfortable doing this. We should skip schema validation until we resolve.

Comment (skshetry) [inline dvc/utils/strictyaml.py:272]: I am not comfortable doing this. I am looking if we can make this lazy.

**Comment by dberenbaum** [comment]
@skshetry Thanks for the feedback. Is there somewhere you have in mind or could find where we could revalidate after resolving all interpolations? Do we need to dump back to yaml in memory?

I took a quick look again but I'm getting confused trying to find where this could happen.

**Comment by dberenbaum** [comment]
@skshetry Thanks for the feedback. Is there somewhere you have in mind or could find where we could revalidate after resolving all interpolations? Do we need to dump back to yaml in memory?

I took a quick look again but I'm getting confused trying to find where this could happen.

> Lazier schema evaluation may end up getting complicated. So I think we should evaluate it eagerly. That’d mean dropping stage-level “vars”.

I'm fine with trying to evaluate as eagerly as we can, but I think we would still have issues matching line numbers and other info to the original yaml since we've already modified it during evaluation.

**Comment by skshetry** [comment]
It probably has to happen here or inside `DataResolver` based on granularity of the schema (f we want it to be lazy):
https://github.com/iterative/dvc/blob/1c394986db1d583a3e269a4352814343679ddf4b/dvc/stage/loader.py#L139

If it has to happen eagerly, we'd need to do resolving/validating somewhere in `ProjectFile.load`.

> I'm fine with trying to evaluate as eagerly as we can, but I think we would still have issues matching line numbers and other info to the original yaml since we've already modified it during evaluation.

We don't modify original yaml data. But they don't have line number anyway, so they are useless.

So, there are two ways we get line information currently:
- If the YAML parsing fails, `ruamel.yaml` itself provides those line numbers. 
- If the schema validation fails, we get the position in terms of keys where it failed, eg: `stages.stage1.cmd`, etc. We keep the plain-text of the dvc.yaml around and when it errors, we try to get the line number for `stages.stage1.cmd`.
  This is done by parsing that plain-text with a round-trip parser, which will have line number information for all keys ([except for leaf values](https://github.com/iterative/dvc/blob/1c394986db1d583a3e269a4352814343679ddf4b/dvc/utils/strictyaml.py#L153-L170)).

So to get line information, we will need to keep track of what we are resolving, and then we can get line number similarly.

File: dvc/stage/loader.py [hunk 5]
```diff
@@ -140,6 +145,8 @@ def __getitem__(self, name):
         except EntryNotFound:
             raise StageNotFound(self.dvcfile, name)  # noqa: B904
 
+        validate(resolved_data, Schema(SCHEMA))
```
Comment (dberenbaum) [inline dvc/stage/loader.py:148]: @skshetry Before I clean up the rest of the PR or tests, is this what you were thinking? We would still need to catch the errors on the first round of validation to get to this point, but at least we then use the existing schema to revalidate the resolved schema.

Comment (dberenbaum) [reply dvc/stage/loader.py:148]: @skshetry ping. WDYT about this approach?

**Comment by dberenbaum** [comment]
@skshetry Can we do this easier now that we have dropped stage-level vars?

**Comment by efiop** [comment]
@skshetry ping

**Comment by skshetry** [comment]
We'll need to relax schema validation (which has been done by @dberenbaum here) and improve error messages. Merging this without that will be a regression in our UI improvement works that was done for parsing.

We have to somehow translate the (lazy) error message to the error location (line number/file etc.). To do that, we have to unify the parsing/parametrization somehow. I'll take over this PR, but will work on it (not immediately though, let's wait for `vars` discussion to conclude).

**Comment by dberenbaum** [comment]
@shcheklein has a request for interpolating `frozen` (same as #6989) in https://iterativeai.slack.com/archives/CB41NAL8H/p1690828656370209?thread_ts=1690236512.298019&cid=CB41NAL8H.

**Comment by skshetry** [comment]
Closing the PR for now as I'm unlikely to get to it and it's been sitting for too long. Let's re-prioritize it separately.
